Mathlib/CategoryTheory/Retract.lean

```diff
@@ -0,0 +1,87 @@
+/-
+Copyright (c) 2024 Jack McKoen. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jack McKoen
+-/
+import Mathlib.CategoryTheory.MorphismProperty.Basic
+
+/-!
+# Retracts
+
+Defines retracts of objects and morphisms.
+
+-/
+
+universe v u
+
+namespace CategoryTheory
+
+variable {C : Type u} [Category.{v} C]
+
+/-- An object `X` is a retract of `Y` if there are morphisms `i : X ⟶ Y` and `r : Y ⟶ X` such
+that `i ≫ r = 𝟙 X`. -/
+class Retract (X Y : C) where
+  /-- `i : X ⟶ Y` -/
+  i : X ⟶ Y
+  /-- `r : Y ⟶ X` -/
+  r : Y ⟶ X
+  /-- `i ≫ r = 𝟙 X` -/
+  retract : i ≫ r = 𝟙 X
+
```

Comment:
Some basic API seems necessary (import `CategoryTheory.EpiMono`):
```suggestion
namespace Retract

attribute [reassoc (attr := simp)] retract

variable {X Y : C} (h : Retract X Y)

@[simps] def splitEpi : SplitEpi h.r where
  section_ := h.i

@[simps] def splitMono : SplitMono h.i where
  retraction := h.r

instance : IsSplitEpi h.r := ⟨⟨h.splitEpi⟩⟩
instance : IsSplitMono h.i := ⟨⟨h.splitMono⟩⟩

end Retract
```
And then, similar definitions and instances for the `_left/_right` variants should be introduced in the `RetractArrow`.